Update tutorial to use autolocking (#12859)
ocaml/dune: doc/tutorials/dune-package-management/locking.md

# Locking Your Dependencies

In the default use-case Dune will automatically determine which packages to
install, by reading the package constrains, determining compatible versions and
installing the dependencies automatically.

For many projects this is a good and acceptable behavior as users often want to
use new versions of their dependencies. However some projects might want to
keep a fixed set of (transitive) dependencies that is only updated manually.

## Create a lock directory manually

Dune resolves dependencies by creating an internal lock directory

> automatically. However, if an manually created lock directory exists in the

Reviewer: At minimum:

```suggestion
automatically. However, if a manually created lock directory exists in the
```

source, Dune will use that one instead. The default name of said lock directory
is `dune.lock`, and it can be created manually:

```
$ dune pkg lock
Solution for dune.lock:
- ocaml.5.2.0
- ocaml-base-compiler.5.2.0
- ocaml-config.3
```

Whenever Dune encounters a `dune.lock` folder, it will use the set of
dependencies defined in the lock. It contains all the metadata about package
names and versions, their dependencies & source locations that are necesary to
build the projects dependencies.

On the next build, Dune will read the stored solver solution from the
`dune.lock` directory, download and build the dependencies and then continue on
building the project as usual.

The lock directory will not be updated until a new lock directory is created,

> by rerunning `dune pkg lock`.

Reviewer: ```suggestion
The lock directory will not be updated until `dune pkg lock` is rerun.
```

> # Locking Your Dependencies

Reviewer: Yeah, I think we should preserve "locking" for the strong sense. You can also fix the dependency version constraints as tightly as you want, but that does not guarantee a lock. I am not even sure we should refer to repo pinning as a way to lock, but that at least gets you the assurance of the same solution space, if you're careful about it.

# Locking Your Dependencies

In the default use-case Dune will automatically determine which packages, their

> versions and their recursive dependencies to install. However, given that new

Reviewer: The first sentence was hard for me to parse. Could something like this make it clearer?

```suggestion
In the default use-case Dune will automatically determine which packages to
install, including solving compatible versions and installing their dependencies
recursively. However, given that new
```

packages are published in the opam-repository occasionally, Dune might pick
different solutions over time.

For many projects this is a good and acceptable behavior as users often want to
use new versions of their dependencies. However some projects might want to
keep a consistent list of dependencies that is only updated manually. This can

> be done in multiple ways.

Reviewer: ```suggestion
be done in two ways.
```

## Create a lock file explicitely

This approach will run the solver with your current package dependency
definitions and create a directory named `dune.lock` which contains the result
of running the solver.

```
$ dune pkg lock
Solution for dune.lock:
- ocaml.5.2.0
- ocaml-base-compiler.5.2.0
- ocaml-config.3
```

Whenever Dune encounters a `dune.lock` folder, it will use the solution defined
in that folder instead of running the solver again. This way the solution will

> only get updated when the user runs the command.

Reviewer: ```suggestion
Whenever Dune encounters a `dune.lock` folder, it will use the set of dependencies defined in the lock. The lock will not be updated until
a user relocks by running `dune pkg lock` again.
```

On the next Dune build, Dune will read the solution from the `dune.lock`
directory, download and build the dependencies and then continue on building
the project as usual.

:::{note}
This approach is similar to using `opam switch export --full --freeze` to
export the configuration of a switch.
:::

> ## Lock the inputs

Reviewer: Maybe let's not overload the term 'lock'? So something like

```suggestion
## Pin the package repositories to a commit
```

Another way to ensure the solution that Dune will pick is to make sure the
inputs to the solver are identical.

::::

::::{dropdown} `dune-workspace`
:icon: file-code

:::{literalinclude} locking/dune-workspace
:language: dune
:::

Dune is instructed to use very specific versions of the repositories to create

> a new solution.

Reviewer: With the rewording I've suggested, I think this becomes unnecessary:

```suggestion
```

::::

On the next build with Dune, Dune will create a new solution by checking out
these specific two repositories and running the solver. It will then download

> and build the dependencies as usual.

Reviewer: ```suggestion
On the next build with Dune, check out these two repositories at the specified
commits and use them for resolving all dependencies. It will then download
and build the dependencies as usual.
```

:::{note}
This approach is similar to pinning OPAM repositories by explicitely setting

> them to know versions using commands like `opam repository add <name> <fixed-url>`.

Reviewer: ```suggestion
them to known versions using commands like `opam repository add <name> <fixed-url>`.
```
